In [1]:
import numpy as np

rng=np.random.default_rng(seed=42)
raw= rng.normal(loc=68,scale=14,size=200)
data= np.clip(np.round(raw,1),0,100)

print(f"Dataset: {len(data)} exam scores")
print(f"First 10 data: {data[:10]}")

#Mean
mean=np.mean(data)
mean_manual=np.sum(data)/len(data)
print(f"Mean:{mean:.4f}")
print(f"Mean manual:{mean_manual}")
print(f"is close :{np.isclose(mean,mean_manual)}")

#Weighted mean

scores  = np.array([85.0, 72.0, 91.0, 68.0])
weights = np.array([0.20, 0.20, 0.30, 0.30])

weighted_mean= np.sum(scores*weights)/np.sum(weights)
print(f"weighted mean:{weighted_mean:.4f}")
print(f"np.average:{np.average(scores,weights=weights):.4f}")
print(f"isclose: {np.isclose(weighted_mean,np.average(scores,weights=weights))}")


Dataset: 200 exam scores
First 10 data: [72.3 53.4 78.5 81.2 40.7 49.8 69.8 63.6 67.8 56.1]
Mean:67.5275
Mean manual:67.5275
is close :True
weighted mean:79.1000
np.average:79.1000
isclose: True


Median = the middle value when data is sorted.
sorted = [55, 60, 65, 70, 72, 74, 78, 82, 85, 88]
                              ↑
                     middle value = (72+74)/2 = 73
                     (even number of values → average of two middle)

odd number of values:
[55, 60, 65, 70, 72, 74, 78, 82, 85]
                    ↑
              middle = 72  (index 4 of 9 values)
              
Mean vs Median — why it matters
Income data:
[30000, 35000, 40000, 45000, 50000, 1000000]
                                      ↑
                                one billionaire

mean   = 200000  ← pulled way up by outlier
median = 42500   ← represents typical person better
Rule of thumb:
mean ≈ median  → symmetric distribution  → use mean
mean > median  → right skewed (rich outliers pulling up) → use median
mean < median  → left skewed → use median

normal distribution   → symmetric by definition  → mean = median

#np.append method
np.append(original_array, new_values, axis)
append( where_to_add , what_to_add , direction )
*It does NOT change the original array
*Default axis=None flattens arrays

In [2]:
median= np.median(data)
#median manual
data_sorted= np.sort(data)
n=len(data_sorted)
if n//2==1:
    median_manual= data_sorted[n//2]
else:
    median_manual= (data_sorted[n//2-1]+data_sorted[n//2])/2
    
print(f"median: {median}")
print(f"median_manual:{median_manual}")
print(f"is close: {np.isclose(median,median_manual)}")

#Outliers
clean= np.array([55, 60, 65, 70, 72, 74, 78, 82, 85, 88])
with_outlier= np.append(clean,[200])
print(f"\nwithout outlier")
print(f"mean: {np.mean(clean):.2f}"
       f"\nmedian:{np.median(clean):.2f}")
print("\nwith outlier")
print(f"mean:{np.mean(with_outlier):.2f}"
      f"\nmedian:{np.median(with_outlier):.2f}")

print(f"\nYour data: mean:{mean:.2f} median: {median:.2f}")
if mean>median:
    print("Slight right-skewed")
elif mean<median:
    print("Slightly left skewed")
else:
    print("symmetric")


median: 67.25
median_manual:67.25
is close: True

without outlier
mean: 72.90
median:73.00

with outlier
mean:84.45
median:74.00

Your data: mean:67.53 median: 67.25
Slight right-skewed


**MODE**
Mode is the most frequently occurring value. 
NumPy has no built in mode function — you implement it manually using np.unique.

arr = [3, 1, 2, 3, 4, 2, 3]

vals, counts = np.unique(arr, return_counts=True)

vals   = [1, 2, 3, 4]   ← unique values sorted
counts = [1, 2, 3, 1]   ← how many times each appears
mode = value with highest count
counts.argmax() = index 2
vals[2] = 3 → mode is 3

**Simple rule**
continuous data  (67.3, 67.4, 67.5...)  → mode is useless
discrete data    (1, 2, 3, 4, 5...)     → mode can be useful
categorical data (A, B, C / yes, no...) → mode is very useful

**Multimodal Distribution:** if multiple values share the highest count → all are modes

In [3]:
rng= np.random.default_rng(seed=42)
integer_scores= rng.integers(50,101,size=50)
vals,counts= np.unique(integer_scores,return_counts=True)
mode_value=vals[np.argmax(counts)]
mode_count= np.max(counts)
print(f"mode:{mode_value} (appears {mode_count} times)")


all_modes = vals[counts == mode_count]
print(f"All modes: {all_modes}")


mode:72 (appears 6 times)
All modes: [72]


## Variance and Standard Deviation

### What they measure
How spread out values are from the mean.
- Low std → values clustered close to mean
- High std → values scattered far from mean

### Formulas
```python
# step by step manually:
deviations = data - np.mean(data)   # distance from mean
sq_dev     = deviations**2          # square to remove negatives
variance   = np.sum(sq_dev) / (n-1) # average squared deviation
std        = np.sqrt(variance)       # back to original units
```

### Population vs Sample
| | Formula | When to use |
|---|---|---|
| Population | divide by n | you have ALL data |
| Sample | divide by n-1 (ddof=1) | you have a sample |

**Rule: almost always use ddof=1 in practice**

Bessel's correction (n-1) compensates for sample mean being
slightly off from true population mean.

### Coefficient of Variation (CV)
Compares spread across datasets with different scales.
```python
cv = (std / mean) * 100
```
- CV < 15%  → low variability
- CV 15-30% → moderate variability  
- CV > 30%  → high variability

**Use case:** comparing spread of exam scores (0-100)
vs salaries (0-100000) on equal footing.

In [4]:
var_pop= np.var(data)
var_sample= np.var(data,ddof=1)
std_pop= np.std(data)
std_sample=np.std(data,ddof=1)
print(f"population variance:{var_pop:.4f}")
print(f"sample variance:{var_sample:.4f}")
print(f"population standard:{std_pop:.4f}")
print(f"sample standard: {std_sample:.4f}")

#manual variance

'''population variance → divide by n
  use when: you have ALL the data (every student in the world)

sample variance → divide by n-1
  use when: you have a SAMPLE (200 students out of millions)'''
        
deviation= data-np.mean(data)
sq_dev = deviation**2
manual_variance= np.sum(sq_dev)/(len(sq_dev)-1)
print(f"\nmanual variance:{manual_variance:.4f}")
print(f"isclose: {np.isclose(manual_variance,var_sample)}")
print(f"Manual Std :{np.sqrt(manual_variance):.4f}")

#Coefficient of variation

cv= (std_sample/mean)*100
print(f"Coefficient of variation:{cv:.2f}%")


population variance:148.4794
sample variance:149.2255
population standard:12.1852
sample standard: 12.2158

manual variance:149.2255
isclose: True
Manual Std :12.2158
Coefficient of variation:18.09%


**Percentiles together**
P0   = minimum value    → 0% below, 100% above
P25  = Q1               → 25% below, 75% above
P50  = median           → 50% below, 50% above  ← you knew this one
P75  = Q3               → 75% below, 25% above
P100 = maximum value    → 100% below, 0% above

sorted scores = [40, 45, 52, 58, 61, 65, 68, 70, 72, 75, 78, 82...]

25th percentile = 58

meaning: 25% of students scored below 58
         75% of students scored above 58
         
**IQR — Interquartile Range**
IQR = Q3 - Q1
    = P75 - P25
IQR represents the middle 50% of data spread

**Real world use**
- P10 → bottom 10% performers → need extra help
- P90 → top 10% performers → excelling
- IQR → typical majority range

In [5]:
# ── Percentiles and quartiles ───────────
percentiles = np.percentile(data, [0, 10, 25, 50, 75, 90, 100])
labels = ['Min (P0)', 'P10', 'Q1 (P25)', 'Median (P50)',
          'Q3 (P75)', 'P90', 'Max (P100)']

print("\nPercentile Summary")
print('-'*25)
print(f"{'Percentile':<16} {'Value':>2}")
print('-'*25)
for label, value in zip(labels,percentiles):
    print(f"{label:<16} {value:<2.2f}")

IQR= np.percentile(data,75)-np.percentile(data,25)
print(f"\nIQR : {IQR:.2f}")

#All values ascending correctly ✓ Median close to 68 ✓


Percentile Summary
-------------------------
Percentile       Value
-------------------------
Min (P0)         38.20
P10              51.40
Q1 (P25)         58.77
Median (P50)     67.25
Q3 (P75)         75.53
P90              82.96
Max (P100)       100.00

IQR : 16.75


**Skewness and Kurtosis**
symmetric (normal):          right skewed:
     *                            *
    ***                          ***
   *****                        *****
  *******                      *******
 *********                    ********.....
55   68   81                 55  68  81  150
     ↑                            ↑        ↑
   center                       center   long right tail
                                          pulling mean right
Skewness
skewness = 0    → perfectly symmetric
skewness > 0    → right tail (positive skew)
skewness < 0    → left tail (negative skew)

|skewness| < 0.5  → approximately symmetric
|skewness| > 1.0  → significantly skewed

Formula for Skewness
**skewness = mean of ((data - mean) / std)**3

Squaring removes negatives (used in variance). Cubing keeps the sign — so left deviations give negative values, right deviations give positive values. The average tells you which side dominates.

**Kurtosis — measuring tail heaviness
kurtosis= np.mean((deviations/std)**4)**
->how fat or thin the tails are compared to a normal distribution.

Formula 

**kurtosis = 0   → same tails as normal (mesokurtic)
kurtosis > 0   → fatter tails → more outliers (leptokurtic)
kurtosis < 0   → thinner tails → fewer outliers (platykurtic)**

Value type	Interpretation
Raw kurtosis > 3	heavier tails
Raw kurtosis = 3	normal tails
Raw kurtosis < 3	lighter tails
Excess kurtosis > 0	heavier tails
Excess kurtosis = 0	normal tails
Excess kurtosis < 0	lighter tails

**Demonstration with three distributions**

1. right skewed  → rng.exponential()
-> classic example: income, house prices, waiting times
-> most values small, few very large values

2. left skewed   → 100 - rng.exponential()
-> classic example: exam scores near max
-> most students score high, few score very low

3. your data     → rng.normal()
->approximately symmetric
->skewness close to 0

In [6]:
deviations= data-np.mean(data)
std= np.std(data,ddof=1)

skewness= np.mean((deviations/std)**3)
kurtosis= np.mean((deviations/std)**4)

print(f"Skewness: {skewness:.4f}")
print(f"Kurtosis: {kurtosis:.4f}")

right_skewed= rng.exponential(scale=5,size=500)
left_skewed= 100-rng.exponential(scale=5,size=500)

for name,dist in [("Right skewed",right_skewed),
                   ("Left skewed",left_skewed),
                   ("your data",data)]:
    dev= dist-np.mean(dist)
    std=np.std(dist,ddof=1)
    sk= np.mean((dev/std)**3)
    ku=np.mean((dev/std)**4)-3
    if sk>0.1:
        direction= "->right_tail"
    elif sk<-0.1:
        direction="->left_tail"
    else:
        direction= symmetric
    print(f"\n{name}")
    print(f"Skewness: {sk:.4f} {direction} ")
    print(f"kurtosis: {ku:.4f}")



Skewness: 0.1804
Kurtosis: 2.6973

Right skewed
Skewness: 1.7765 ->right_tail 
kurtosis: 3.9671

Left skewed
Skewness: -2.1162 ->left_tail 
kurtosis: 6.4939

your data
Skewness: 0.1804 ->right_tail 
kurtosis: -0.3027


In [7]:
#Method 1 - Z score method
data_o= np.append(data, [150, 160, -10, 155])
mean_o= np.mean(data_o)
std_o=np.std(data_o)

z_scores= (data_o-mean_o)/std
outlier_mask_z=np.abs(z_scores)>3
print("\nZ Score (threshold=3)")
print(f"No of outliers: {np.sum(outlier_mask_z)}")
print(f"outlier values:{data_o[outlier_mask_z]}")
print(f"z_scores of outliers: {z_scores[outlier_mask_z]}")

#Method2 IQR method

q1_o= np.percentile(data_o,25)
q3_o=np.percentile(data_o,75)
iqr_o= q3_o-q1_o

#mild outliers
lower_fence= q1_o-(1.5*iqr_o)
upper_fence= q3_o+(1.5*iqr_o)

#extreme outliers
extreme_lower= q1_o-(3*iqr_o)
extreme_upper= q3_o+(3*iqr_o)

print("\nIQR METHOD")
print(f"Q1:{q1_o:.2f} Q3: {q3_o:.2f} IQR: {iqr_o:.2f}")
print(f"lower fence: {lower_fence:.2f}")
print(f"Upper fence: {upper_fence:.2f}")
print(f"Extreme lower:{extreme_lower:.2f}")
print(f"Extreme upper:{extreme_upper:.2f}")

mild_outliers= data_o[(data_o<lower_fence)|(data_o>upper_fence)]
print(f"length of mild outliers:{len(mild_outliers)} outlier_values:{mild_outliers}")
extreme_outliers=data_o[(data_o<extreme_lower)|(data_o>extreme_upper)]
print(f"length of extreme outliers: {len(extreme_outliers)} outlier_values: {extreme_outliers}")

#Method 3 Modified Z Score
def modified_z_score(arr):
    median_o= np.median(arr)
    mad= np.median(np.abs(arr-median_o))
    modified_z= 0.6745*((arr-median_o)/mad)
    return modified_z
mod_z=modified_z_score(data_o)
outlier_mask_modz= np.abs(mod_z)>3.5 
outliers= data_o[outlier_mask_modz]
print("\nModified Z Score Method")
print("Modified z score (threshold=3.5)")
print(f"No of outliers: {len(outliers)}")
print(f"outlier values: {outliers}")

# ── Compare all three methods ────────────────────────────────
print("\nMethod Comparison:")
print(f"{'Method':<25} {'Outliers Found':>15} {'Values'}")
print("─" * 65)

z_out   = data_o[np.abs(z_scores) > 3]
iqr_out = data_o[
    (data_o < lower_fence) | (data_o > upper_fence)
]
mz_out  = data_o[np.abs(mod_z) > 3.5]

for name, out in [("Z-score (>3)", z_out),
                  ("IQR (1.5×IQR)", iqr_out),
                  ("Modified Z (>3.5)", mz_out)]:
    vals = sorted(out.tolist())
    print(f"{name:<25} {len(out):>15}      {vals}")



Z Score (threshold=3)
No of outliers: 4
outlier values:[150. 160. -10. 155.]
z_scores of outliers: [ 6.67711033  7.49572297 -6.42069195  7.08641665]

IQR METHOD
Q1:58.77 Q3: 76.30 IQR: 17.52
lower fence: 32.49
Upper fence: 102.59
Extreme lower:6.20
Extreme upper:128.88
length of mild outliers:4 outlier_values:[150. 160. -10. 155.]
length of extreme outliers: 4 outlier_values: [150. 160. -10. 155.]

Modified Z Score Method
Modified z score (threshold=3.5)
No of outliers: 4
outlier values: [150. 160. -10. 155.]

Method Comparison:
Method                     Outliers Found Values
─────────────────────────────────────────────────────────────────
Z-score (>3)                            4      [-10.0, 150.0, 155.0, 160.0]
IQR (1.5×IQR)                           4      [-10.0, 150.0, 155.0, 160.0]
Modified Z (>3.5)                       4      [-10.0, 150.0, 155.0, 160.0]


In [8]:
#Practise Drill
rng= np.random.default_rng(seed=42)
database= rng.normal(loc=50,scale=10,size=100)
database= np.clip(np.round(database,1),0,100)
mean=np.mean(database)
deviations= database-mean
std=np.std(database,ddof=1)
sk= np.mean((deviations/std)**3)
ku= np.mean((deviations/std)**4)

print(f"mean:     {mean:.4f}")
print(f"std:      {std:.4f}")
print(f"skewness: {sk:.4f}")
print(f"kurtosis: {ku:.4f}")

database_o= np.append(database,[200, -50, 300])
mean_o=np.mean(database_o)
std_o=np.std(database_o,ddof=1)
#outliers
#z_score method
z_score= (database_o-mean_o)/std_o
outlier_mask_z= np.abs(z_score)>3
print(f"Outliers-z_score:{database_o[outlier_mask_z]}")

#IQR method
Q1=np.percentile(database_o,25)
Q3=np.percentile(database_o,75)
IQR= Q3-Q1
lower_fence=Q1 -(1.5*IQR)
upper_fence=Q3+(1.5*IQR)
outlier_mask_IQR= (database_o<lower_fence)|(database_o>upper_fence)
print(f"IQR method outliers: {database_o[outlier_mask_IQR]}")

mean:     49.4990
std:      7.7682
skewness: -0.1369
kurtosis: 2.6729
Outliers-z_score:[200. -50. 300.]
IQR method outliers: [200. -50. 300.]


**Correlation, Covariance & Histogram Analysis**

**Part 1-Covariance and correlation**
-> Covariance measures how two variables move together.

study_hours and exam_score:
  student studies 8 hrs → scores 85
  student studies 2 hrs → scores 40
  when study goes up, score goes up → POSITIVE covariance

study_hours and sleep_hours:
  student studies 8 hrs → sleeps 5 hrs
  student studies 2 hrs → sleeps 8 hrs
  when study goes up, sleep goes down → NEGATIVE covariance

study_hours and random_var:
  no relationship whatsoever → NEAR ZERO covariance

**Formula- 
cov(X, Y) = sum((xi - x_mean) × (yi - y_mean)) / (n-1)
**

(xi - x_mean) → how far each x value is from x mean
(yi - y_mean) → how far each y value is from y mean

multiply them together:
  both above mean → positive × positive = positive
  both below mean → negative × negative = positive
  one above, one below → positive × negative = negative

sum all products → tells you which direction dominates
divide by n-1   → sample  positive × negative = negative

sum all products → tells you which direction dominates
divide by n-1   → sample covariance

**Pearson Correlation — fixes the scale problem**

Problem with covariance — it depends on scale
study hours range: 1-10
exam score range:  0-100

cov(study, exam) = 40.5   ← large number
covariance

**Pearson Correlation — fixes the scale problem**

Problem with covariance — it depends on scale
study hours range: 1-10
exam score range:  0-100

cov(study, exam) = 40.5   ← large number

study hours range: 1-10
sleep hours range: 3-10

cov(study, sleep) = -0.8  ← small number
**r = cov(X, Y) / (std(X) × std(Y))**

->np.corrcoef is not a separate method — it is just NumPy's built-in function that computes Pearson correlation for you.

->Dividing by both standard deviations removes the units. Result is always between -1 and +1:

r = +1.0  → perfect positive  (study doubles, score always doubles)
r = +0.8  → very strong positive
r = +0.5  → moderate positive
r =  0.0  → no relationship
r = -0.5  → moderate negative
r = -1.0  → perfect negative

Interpretation guide:
|r| >= 0.8  → very strong
|r| >= 0.6  → strong
|r| >= 0.4  → moderate
|r| >= 0.2  → weak
|r| <  0.2  → negligible

1. example
exam_score = 7.5 * study_hours + rng.normal(0, 7, n)
signal strength  = 7.5 × (range of study hours)
                 = 7.5 × (10-1) = 67.5 marks range (75-7.5)
noise strength   = std 7 marks ( noise only disturbs them by ±7 marks. )
signal >> noise → strong relationship → r will be high positive (~0.9)

2. example
sleep_hours = -0.3 * study_hours + rng.normal(7, 1, n)
signal strength = 0.3 × (range of study hours)
                = 0.3 × (10-1) = 2.7 hours range
noise strength  = std 1 hour
signal is only ~3× bigger than noise → weak relationship → r will be small negative (~-0.3)

**General Rule : 
large coefficient + small noise → strong correlation
small coefficient + large noise → weak correlation
**



In [9]:

rng= np.random.default_rng(seed=42)
n=100
study_hrs= rng.uniform(1,10,n)#every value b/w 1 and 10 has equal chance of appearing
exam_score= 7.5 * study_hrs+ rng.normal(0,7,n)
exam_score=np.clip(exam_score,0,100)

sleep_hrs= -0.3 * study_hrs + rng.normal(7,1,n)
sleep_hrs= np.clip(sleep_hrs,3,10)
random= rng.normal(50,10,n)


def manual_cov(x,y,ddof=1):
    n=len(x)
    x_mean= np.mean(x)
    y_mean= np.mean(y)
    x_dev= x-x_mean
    y_dev= y-y_mean
    return np.sum(x_dev*y_dev)/(n-ddof)

cov_study_exam= manual_cov(study_hrs,exam_score)
cov_study_sleep= manual_cov(study_hrs,sleep_hrs)
cov_study_random= manual_cov(study_hrs,random)

print("Covariances:")
print(f"  study vs exam:   {cov_study_exam:.4f}  (positive → study more, score higher)")
print(f"  study vs sleep:  {cov_study_sleep:.4f} (negative → study more, sleep less)")
print(f"  study vs random: {cov_study_random:.4f}  (near zero → no relationship)")
#the numbers are in different scales so direct comparison is hard. This is exactly why we need correlation next.

#pearson correlation

def pearson_r(x,y):
    cov_xy= manual_cov(x,y)
    std_x= np.std(x,ddof=1)
    std_y=np.std(y,ddof=1)
    return cov_xy/((std_x)*(std_y))

r_study_exam= pearson_r(study_hrs,exam_score)
r_study_sleep= pearson_r(study_hrs,sleep_hrs)
r_study_random=pearson_r(study_hrs,random)

print("Pearson correlations: ")
print(f"Study vs Exam: {r_study_exam:.4f}")
print(f"Study vs Sleep: {r_study_sleep:.4f}")
print(f"Study vs Random: {r_study_random:.4f}")
print(" ")
for name , r in [("Study vs Exam", r_study_exam),("Study vs Sleep",r_study_sleep),
                 ("Study vs Random",r_study_random)]:
    if   np.abs(r)>=0.8: Strength="Very strong"
    elif np.abs(r)>=0.6: Strength= "Strong"
    elif np.abs(r)>=0.4: Strength= "Moderate"
    elif np.abs(r)>=0.2: Strength|= "Weak"
    else:                Strength="Negligible"
    direction= "Positive" if r>0 else  "Negative"
    print(f"{name}: {Strength}  {direction} correlation")  
    
## ── Full correlation matrix using np.corrcoef ────────────────

data_matrix= np.column_stack((study_hrs,exam_score,sleep_hrs,random))#(100,4)
correl_matrix= np.corrcoef(data_matrix.T)#(4,4)
names=["Study Hrs", "Exam Score", "Sleep Hrs", "Random"]
print(f"\n{' ':>12}",end="")
for name in names:
    print(f"{name:>12}",end="")
print()
for i , row_name in enumerate(names):
    print(f"{row_name:<12}",end="")
    for j in range(len(names)):
        val= correl_matrix[i,j]
        marker= "⭐" if np.abs(val)>=0.5 and i!=j else " "
        print(f"{val:>10.2f} {marker}",end="")
    print()
    
# ── Covariance matrix using np.cov ───────────────────────────
cov_matrix=np.cov(data_matrix.T)#(4,4)
print(f"\nCo variance matrix: Diagonal(variances):")
print(f"{cov_matrix.round(2)}")
print(f"Diagonal variances: {np.diag(cov_matrix).round(2)}")

# Verify: correlation = cov / (std_i * std_j)
std_vector= np.std(data_matrix,axis=0,ddof=1)
std_outer= np.outer(std_vector,std_vector)#(4x4)
corr_from_cov= cov_matrix/std_outer
print("Correlation from cov matrix:")
print(f"{corr_from_cov.round(2)}")
print(f"Match correl  matrixes: {np.allclose(correl_matrix,corr_from_cov)}")

# ── Finding strongest correlations automatically ─────────────
def find_strong_correlations(correl_matrix,names,threshold=0.3):
    correl_matrix=correl_matrix.round(2)
    n=len(names)
    strong_pairs=[]
    for i in range(n):
        for j in range(i+1,n):
            r=correl_matrix[i,j]
            if np.abs(r)>threshold:
                strong_pairs.append((names[i],names[j],r))
    strong_pairs.sort(key=lambda x: np.abs(x[2]),reverse=True)#Descending order sorting
    return strong_pairs
strong_pairs= find_strong_correlations(correl_matrix,names)
print(strong_pairs)

print("\nSignificant correlations (|r| > 0.2):")
print(f"{'Feature A':<14} {'Feature B':<14} {'r':>8}  {'Strength'}")
print("─" * 50)
for pairs in strong_pairs:
    r=np.abs(pairs[2])
    strength= "⭐⭐⭐" if r>0.8 else("⭐⭐" if r>0.6 else "⭐") 
    print(f"{pairs[0]:<14} {pairs[1]:<14} {pairs[2]:>8} {strength}")
    
                
    



    



Covariances:
  study vs exam:   45.8436  (positive → study more, score higher)
  study vs sleep:  -2.2104 (negative → study more, sleep less)
  study vs random: -0.6177  (near zero → no relationship)
Pearson correlations: 
Study vs Exam: 0.9378
Study vs Sleep: -0.6787
Study vs Random: -0.0244
 
Study vs Exam: Very strong  Positive correlation
Study vs Sleep: Strong  Negative correlation
Study vs Random: Negligible  Negative correlation

               Study Hrs  Exam Score   Sleep Hrs      Random
Study Hrs         1.00        0.94 ⭐     -0.68 ⭐     -0.02  
Exam Score        0.94 ⭐      1.00       -0.63 ⭐     -0.07  
Sleep Hrs        -0.68 ⭐     -0.63 ⭐      1.00       -0.01  
Random           -0.02       -0.07       -0.01        1.00  

Co variance matrix: Diagonal(variances):
[[ 6.0600e+00  4.5840e+01 -2.2100e+00 -6.2000e-01]
 [ 4.5840e+01  3.9402e+02 -1.6600e+01 -1.4140e+01]
 [-2.2100e+00 -1.6600e+01  1.7500e+00 -9.0000e-02]
 [-6.2000e-01 -1.4140e+01 -9.0000e-02  1.0529e+02]]
Diagona

**Histogram**
-> np.histogram
counts, bin_edges = np.histogram(exam_score, bins=10)
It divides the data range into bins equal width buckets and counts how many values fall in each bucket.

Example with bins=5:
exam scores range: 0 to 100
bin width = 100/5 = 20

bin 1: 0-20   → count how many scores between 0 and 20
bin 2: 20-40  → count how many scores between 20 and 40
bin 3: 40-60  → count how many scores between 40 and 60
bin 4: 60-80  → count how many scores between 60 and 80
bin 5: 80-100 → count how many scores between 80 and 100
Output:
bin_edges = [0, 20, 40, 60, 80, 100]   ← 6 edges for 5 bins
counts    = [2,  5, 35, 45, 13]        ← how many in each bin
            ↑   ↑   ↑   ↑   ↑
           few  few many most some
           
Custom bins — grade boundaries
pythoncustom_bins = [0, 40, 50, 60, 70, 80, 90, 100]

**Density**

Raw counts depend on sample size:
class A: 200 students → bin 60-70 has count=45
class B:  50 students → bin 60-70 has count=11
bar heights are completely different
even if both classes have same score distribution
→ cannot compare directly

### What density does

Converts raw counts into a value where **total area = 1.0**
Formula:
**density = count / (total_count × bin_width)**
Example:
bin 60-70, count=45, total=200, bin_width=10
density = 45 / (200 × 10) = 0.0225

### Why area = 1.0 matters
**area of one bar = density × bin_width**
= 0.0225 × 10
= 0.225
meaning: 22.5% of students scored between 60-70
total area = sum of all bars = 1.0 = 100%
→ behaves like a probability distribution

### Verifying area = 1.0

```python
density, edges = np.histogram(data, bins=20, density=True)

# np.diff(edges) gives width of each bin
bin_widths = np.diff(edges)   # e.g. [5.0, 5.0, 5.0 ...]

# area = density × width for each bin
total_area = np.sum(density * bin_widths)
print(total_area)   # always exactly 1.0
```

### Visual comparison
Raw counts (misleading when comparing):
class A (n=200):        class B (n=50):
45|  ██                 11|  ██
30|████                  7| ███
20|████                  5|████
40 60 80 100            40 60 80 100
bars look different → hard to compare shapes
Density (fair comparison):
class A:                class B:
0.023| ██             0.022| ██
0.015|████            0.014|███
0.010|████            0.010|████
40 60 80 100          40 60 80 100
bars look similar → shapes clearly comparable




In [10]:
#Histogram
counts,bin_edges= np.histogram(exam_score,bins=10)#exam score range: 1-100
print("Histogram of exam scores:")
print(f"Edges : {bin_edges.round(2)}")
print(f"Count: {counts}")
print(f"Total: {np.sum(counts)} (should be equal to n=100)")

bin_center= (bin_edges[:-1]+bin_edges[1:])/2 #Gives centres of all bins at once (leftedges +right edges )/2
print(f"{'Score Range':<15} {'count':<10} {'Bar'}")
for left,right,count in zip(bin_edges[:-1],bin_edges[1:],counts):
    bar= '◼️' * count
    print(f"{left:5.1f}-{right:5.1f}     {count:<10} {bar}")
#Density
density,bin_edges= np.histogram(exam_score,bins=20,density=True)
bin_widths= np.diff(bin_edges)
total_area= np.sum(density * bin_widths)
print(f"\nWith density=True:")
print(f"Area under histogram: {total_area:.1f}")

#Custom bins
custom_bins = [0, 40, 50, 60, 70, 80, 90, 100]
grade_counts,grade_edges= np.histogram(exam_score,bins= custom_bins)
grade_labels = ['< 40 (F)', '40–50 (D)', '50–60 (C)', '60–70 (C+)', '70–80 (B)', '80–90 (A-)', '90–100 (A)']
total= np.sum(grade_counts)
print("Grade Distribution")
for label,count in zip(grade_labels,grade_counts):
    pct= (count/total) *100
    bar= '◼️'* int(pct)
    print(f"{label:<12}  {count}  ({pct:4.1f}%)   {bar}")
    




Histogram of exam scores:
Edges : [ 5.97 13.79 21.6  29.41 37.23 45.04 52.85 60.66 68.48 76.29 84.1 ]
Count: [11 11 13 11  8 13 15 12  4  2]
Total: 100 (should be equal to n=100)
Score Range     count      Bar
  6.0- 13.8     11         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 13.8- 21.6     11         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 21.6- 29.4     13         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 29.4- 37.2     11         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 37.2- 45.0     8          ◼️◼️◼️◼️◼️◼️◼️◼️
 45.0- 52.9     13         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 52.9- 60.7     15         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 60.7- 68.5     12         ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
 68.5- 76.3     4          ◼️◼️◼️◼️
 76.3- 84.1     2          ◼️◼️

With density=True:
Area under histogram: 1.0
Grade Distribution
< 40 (F)      51  (51.0%)   ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
40–50 (D)     11  (11.0%)   ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️
50–60 (C)     18  (18.0%)   ◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼️◼

Bootstrapping + Permutation Tests

## Part 1 — Bootstrapping

### The problem it solves
You have one sample of 50 scores and compute mean = 74.3.
How reliable is this estimate? Would a different 50 students give a very different mean?
Bootstrapping answers this using only the data you already have.

---

### Core idea — resampling with replacement
Pick n values randomly from your data, putting each back before the next pick.
Same value can appear multiple times. Same size as original.
original = [65, 72, 81, 58, 90]   n=5
resample 1 = [72, 72, 58, 90, 65]  → mean=71.4  (72 repeated)
resample 2 = [81, 90, 90, 74, 58]  → mean=78.6  (90 repeated)
resample 3 = [58, 65, 72, 81, 83]  → mean=71.8  (different mix)
...repeat 10,000 times...

**Why with replacement?**
Without replacement → same values every time → same mean every time → useless
With replacement → different values each time → different means → useful distribution

---

### Steps
```python
1. resample WITH replacement, same size n → 10,000 times
2. compute statistic (mean/median/anything) on each resample
3. sort the 10,000 statistics
4. cut 2.5% from each tail → 95% confidence interval
```

```python

bootstrap_samples = rng.choice(data, size=(10000, n), replace=True)
bootstrap_stats   = np.apply_along_axis(np.mean, axis=1, arr=bootstrap_samples)
ci_lower = np.percentile(bootstrap_stats, 2.5)
ci_upper = np.percentile(bootstrap_stats, 97.5)**
```

---

### Confidence interval
sort 10,000 bootstrap means:
[68.1 ... 71.3 ......................... 77.1 ... 79.4]
↑                               ↑
2.5th percentile            97.5th percentile
= ci_lower                  = ci_upper
95% CI = [71.3, 77.1]

---

### What confidence level means

| What people think | What it actually means |
|---|---|
| 95% chance true mean is in interval | If 100 researchers each built a CI, 95 would contain the true mean |
| interval moves | interval is fixed once computed — the TRUE MEAN is fixed |
true mean = 72.0 (fixed, unknown)
researcher 1: CI = [69.1, 78.3]  ✓ contains 72
researcher 2: CI = [70.2, 79.8]  ✓ contains 72
researcher 3: CI = [73.5, 82.1]  ✗ missed (bad luck)
...95 out of 100 intervals catch it

**Think of it as a fishing net:**
You cast a net. 95% confidence means your net catches the true value 95% of the time.

---

### Choosing confidence level
90% CI → alpha=5.0 → cut at 5.0 and 95.0   → narrower, less sure
95% CI → alpha=2.5 → cut at 2.5 and 97.5   → standard choice
99% CI → alpha=0.5 → cut at 0.5 and 99.5   → wider, more sure
**formula: alpha = (100 - confidence) / 2**

Higher confidence = cut less from tails = wider interval = more certain of catching true value.

---

### Works for ANY statistic — that is the power
analytical CI for median     → complex formula
analytical CI for correlation → even more complex
bootstrap CI for both         → exact same 5 lines of code
bootstrap_ci(sample, np.mean)    → CI for mean
bootstrap_ci(sample, np.median)  → CI for median
bootstrap_ci(sample, np.max)     → CI for maximum

---

### Effect of sample size
n=10  → CI width=12.4  → very uncertain
n=50  → CI width= 5.8  → more certain
n=200 → CI width= 2.9  → quite certain
n=500 → CI width= 1.8  → very certain
more data → narrower CI → more reliable estimate
theoretical: std_error = std / sqrt(n)
bootstrap agrees with this formula ← validates the approach

# Replace=True
rng.choice picks random values from data. The replace=True argument tells it to put each value back before picking the next one.
data = [65, 72, 81, 58, 90]

replace=False (without replacement):
pick 65 → remove it → [72, 81, 58, 90]
pick 72 → remove it → [81, 58, 90]
pick 81 → remove it → [58, 90]
pick 58 → remove it → [90]
pick 90 → done
result = [65, 72, 81, 58, 90] ← just a shuffle, same values always

replace=True (with replacement):
pick 72 → put it back → [65, 72, 81, 58, 90]
pick 72 → put it back → [65, 72, 81, 58, 90]  ← can pick 72 again
pick 58 → put it back → [65, 72, 81, 58, 90]
pick 90 → put it back → [65, 72, 81, 58, 90]
pick 65 → put it back → done
result = [72, 72, 58, 90, 65] ← 72 repeated, 81 missing
Without replace=True every resample would just be a shuffle of the original — same mean every time — completely useless for bootstrapping.
replace=True is what creates the variation between resamples that makes bootstrapping work.



In [11]:
rng= np.random.default_rng(seed=42)
sample= rng.normal(loc=70,scale=15,size=50)
np.clip(sample,0,100)
print(f"Sample Size:{len(sample)}")
print(f"Sample mean:{np.mean(sample):.4f}")
print(f"Sample std:{np.std(sample,ddof=1):.4f}")

# ── Bootstrap implementation ─────────────────────────────────
def bootstrap_ci(data,statistic_fn,n_bootstrap=10000,confidence=95):
    bootstrap_samples= rng.choice(data,size=(n_bootstrap,len(data)),replace=True)#(10,000,50)
    bootstrap_stats= np.apply_along_axis(statistic_fn,axis=1,arr= bootstrap_samples)
    alpha= (100-confidence)/2 # we cut alpha from head and tail
    ci_lower= np.percentile(bootstrap_stats,alpha)
    ci_upper= np.percentile(bootstrap_stats,100-alpha)
    
    
    return {
        'point_estimate':          round(float(statistic_fn(data)), 6),
        'ci_lower':                round(float(ci_lower), 6), # converts number from numpy scalar to python float which is feasible for round function
        'ci_upper':                round(float(ci_upper), 6),
        'bootstrap_std':           round(float(np.std(bootstrap_stats)), 6),
        'bootstrap_distribution':  bootstrap_stats,
        'confidence':              confidence
    }
# ── Bootstrap CI for multiple statistics ────────────────────
statistics= {
             "mean":np.mean,
             "median":np.median,
             "Std Dev": lambda x: np.std(x,ddof=1),
             "Q1(25%)": lambda x: np.percentile(x,25),
             "Q3 (75%)": lambda x:np.percentile(x,75),
             "max"     : np.max,
             
}
print("Bootstrap 95% Confidence Intervals")
print(f"{'Statistic':<14} {'Estimate':>10} {'CI Lower':>10} "
      f"{'CI Upper':>10} {'Width':>8}")
print("─" * 56)

for stat_name,stat_fn in statistics.items():
    result= bootstrap_ci(sample,stat_fn)
    width=result["ci_upper"] -result["ci_lower"]
    print(f"{stat_name:<14} {result['point_estimate']:>10} {result['ci_lower']:>10}",
          f"{result['ci_upper']} {width:>8.2f}")  

Sample Size:50
Sample mean:71.3682
Sample std:11.5242
Bootstrap 95% Confidence Intervals
Statistic        Estimate   CI Lower   CI Upper    Width
────────────────────────────────────────────────────────
mean            71.368165  68.207814 74.504004     6.30
median          72.598967  68.290788 76.190989     7.90
Std Dev         11.524158   9.116758 13.764594     4.65
Q1(25%)         63.657256  57.672939 69.251111    11.58
Q3 (75%)        79.629092  75.531262 82.720294     7.19
max            102.124714  86.908618 102.124714    15.22


Standard error:
It measures how spread out the sample means are from the true population mean.

**Formula : std_error = std / sqrt(n)**

population: all students in India (millions)
true mean exam score = 72.0

you collect sample 1 (n=50) → mean = 71.3
you collect sample 2 (n=50) → mean = 73.1
you collect sample 3 (n=50) → mean = 70.8
you collect sample 4 (n=50) → mean = 72.5
...

these sample means vary around 72.0
how much they vary = standard error

1. CI width nearly halves when n quadruples
n=10  → 22.74
n=50  → 7.33   (5× more data → width dropped by 3×)
n=200 → 4.56   (4× more data → width dropped by ~2×)
n=500 → 2.56

This matches the formula std_error = std/sqrt(n) — to halve the error you need 4× more data (because sqrt(4)=2).

2. Bootstrap CI width ≈ 4 × std_error
n=50:  CI width = 7.33,  std_error = 1.91,  ratio = 7.33/1.91 = 3.84 ≈ 4 ✓
n=100: CI width = 6.08,  std_error = 1.54,  ratio = 6.08/1.54 = 3.95 ≈ 4 ✓
n=200: CI width = 4.56,  std_error = 1.15,  ratio = 4.56/1.15 = 3.97 ≈ 4 ✓
Bootstrap agrees with mathematical theory ✓ — this validates the bootstrap approach.

In [12]:
# ── Effect of sample size on CI width ───────────────────────
# The larger the sample, the narrower the CI (more certainty)
print("Effect of sample size on CI width (for mean)")
print(f"{'Sample_size':>12} {'CI Width':>10} {'Std Error':>12}")
print("-"*36)

for n_size in [10, 25, 50, 100, 200, 500]:
    sub_sample= rng.normal(loc=70,scale=15,size=n_size)
    result=bootstrap_ci(sub_sample,np.mean,n_bootstrap=5000)
    width= result["ci_upper"]-result["ci_lower"]
    std_error= np.std(sub_sample,ddof=1)/np.sqrt(n_size)
    print(f"{n_size:>12} {width:>10.2f} {std_error:>12.2f}")
    


Effect of sample size on CI width (for mean)
 Sample_size   CI Width    Std Error
------------------------------------
          10      14.97         4.09
          25      11.69         3.01
          50       8.11         2.10
         100       6.00         1.53
         200       4.12         1.06
         500       2.65         0.69


In [13]:
# ── Bootstrap for a more complex statistic: correlation ─────
rng= np.random.default_rng(seed=42)
study= rng.uniform(1,10,60)
score= 7*study+rng.normal(0,8,60)
score=np.clip(score,0,100)

def bootstrap_correlation(x,y,n_bootstrap=10000):
    rng= np.random.default_rng(seed=42)
    paired=np.column_stack((x,y))#(60,2)
    n=len(x)
    bootstrap_samples=[]
    for _ in range(n_bootstrap):
        indx= rng.integers(0,n,size=n) #(60)
        resample= paired[indx]#(60,2)
        bootstrap_samples.append(pearson_r(resample[:,0],resample[:,1]))
    bootstrap_samples= np.array(bootstrap_samples)
    return {"r": pearson_r(x,y).round(4),
            "ci_lowerfence": np.percentile(bootstrap_samples,2.5),
            "ci_upperfence": np.percentile(bootstrap_samples,97.5)}
        
        
result=bootstrap_correlation(study,score)
print("Bootstrap CI for pearson r")
print(f"r: {result['r']}")
print(f"95% CI :[{result['ci_lowerfence']:.4f},{result['ci_upperfence']:.4f}]")



Bootstrap CI for pearson r
r: 0.9485
95% CI :[0.9263,0.9658]


Session 3 — Part 2: Permutation Test

Before any code — the concept from scratch
You already understand bootstrapping — you had one sample and wanted to know how uncertain your estimate was. Permutation tests answer a completely different question.
The question: You have two groups. Group A got a new teaching method (mean = 74). Group B got the old method (mean = 68). The difference is 6 points. Is this real, or just luck?
Think about it this way. Imagine you had 80 students and you randomly split them into two groups of 40 — no teaching difference, just random assignment. Sometimes group A would score higher just by chance, sometimes group B would. The question is: how often would you get a 6-point gap just from random splitting?
If random splitting gives 6-point gap often  → our result is nothing special → not significant
If random splitting almost never gives 6 points → our result is rare → real effect → significant
That is the entire logic. Now here is how you simulate it.

The 5-step process
Step 1 → combine both groups into one pool of 80 students
          (assumption: if H0 is true, they all come from same population)

Step 2 → randomly shuffle the 80 students

Step 3 → split first 40 as "fake treatment", last 40 as "fake control"

Step 4 → compute the difference of means for this random split

Step 5 → repeat 10,000 times → collect 10,000 random differences
          count how many are >= our observed 6 points
          that fraction = p-value

H0 — null hypothesis in plain English
H0 = "the teaching method makes no difference"
   = "both groups come from the same population"
   = "the labels treatment/control are meaningless"

If H0 is true → shuffling labels randomly should
                produce similar differences regularly

If H0 is false → our observed difference is RARE
                 under random shuffling → real effect

P-value — what the number actually means
p-value = 0.003

Plain English:
"If the teaching method truly made no difference,
there is only a 0.3% chance of seeing a 6-point gap
just by random luck"

p-value = "probability of seeing this result if H0 were true"

0.3% is tiny → very unlikely to be luck → real effect

─────────────────────────────────────────
Standard threshold: α = 0.05

p < 0.05 → less than 5% chance it is luck → SIGNIFICANT
p > 0.05 → more than 5% chance it is luck → NOT significant

if p < 0.05 → evidence strong enough → reject H0 → method works
if p > 0.05 → evidence weak → fail to reject H0 → not proven

Two-sided vs one-sided
Two-sided: "Is there ANY difference between the groups?"
→ count shuffles where |random diff| >= |observed diff|
→ counts both +6 and -6 as equally extreme

One-sided: "Is treatment specifically BETTER than control?"
→ count shuffles where random diff >= observed diff
→ only counts improvements, ignores drops

Use two-sided when you had no prior prediction of direction.
Use one-sided when you specifically predicted which group
would score higher BEFORE collecting the data.

One-sided p ≈ half of two-sided p — because you are only
counting one tail of the distribution instead of both.

In [14]:
import numpy as np

rng= np.random.default_rng(seed=42)
n=40

control= rng.normal(loc= 68,scale=12,size=n)
treatment= rng.normal(loc=74,scale=12,size=n)
control_mean= np.mean(control)
treatment_mean= np.mean(treatment)
print(f"Control mean: {control_mean:.4f}")
print(f"Treatment mean: {treatment_mean:.4f}")

#Observed difference
observed_diff= treatment_mean-control_mean
print(f"Observed difference: {observed_diff:.4f}")

# combine both into one pool
combined= np.concatenate([control,treatment])
print(f"length of combined: {len(combined)}")

#Permutation loop
n_permutations=10000
perm_diff= np.zeros(n_permutations)

for i in range(n_permutations):
    shuffled=rng.permutation(combined) # randomly reorder all 80
    fake_treatment= shuffled[:n] # group1 first 40
    fake_control= shuffled[n:]
    perm_diff[i]= np.mean(fake_treatment)-np.mean(fake_control)
print(f"Mean of perm_diff: {np.mean(perm_diff):.4f}") # ~ 0
print(f"Std of perm_diff: {np.std(perm_diff):.4f}")
print(f"Min of perm_diff: {np.min(perm_diff):.4f}")
print(f"Min of perm_diff: {np.max(perm_diff):.4f}")

'''np.mean(perm_diffs) should be very close to 0. This makes sense — if groups are randomly assigned, sometimes 
fake_treatment is higher, sometimes fake_control is higher. These cancel out on average.'''

#Compute p value
#two -sided p value
p_value_two= np.mean(np.abs(perm_diff)>= np.abs(observed_diff) )

#one_sided p value
p_value_one= np.mean(perm_diff>= observed_diff)

print(f"Observed difference:    {observed_diff:.4f}")
print(f"P-value (two-sided):    {p_value_two:.4f}")
print(f"P-value (one-sided):    {p_value_one:.4f}")
print(f"Significant at α=0.05:  {p_value_two < 0.05}")
    
#where does your observed diff sit in the distribution:
pct_lower=  np.percentile(perm_diff,2.5).round(4)
pct_upper= np.percentile(perm_diff,97.5).round(4)

print(f"\n95% of random diffs fall between: {[pct_lower,pct_upper]}")
print(f"Observed difference : {observed_diff:.4f}")
if observed_diff>pct_upper or observed_diff<pct_lower:
    print(f"→ OUTSIDE 95% range → SIGNIFICANT ✓")
else:
    print("→ INSIDE 95% range  → not significant")




Control mean: 68.4592
Treatment mean: 74.1581
Observed difference: 5.6989
length of combined: 80
Mean of perm_diff: -0.0036
Std of perm_diff: 2.1735
Min of perm_diff: -8.9248
Min of perm_diff: 8.2794
Observed difference:    5.6989
P-value (two-sided):    0.0082
P-value (one-sided):    0.0046
Significant at α=0.05:  True

95% of random diffs fall between: [-4.2062, 4.2964]
Observed difference : 5.6989
→ OUTSIDE 95% range → SIGNIFICANT ✓


markdown## Session 4 Project: Clinical Trial Analyser — What Each Output Means

---

### Step 1 — Data Generation
Control   reduction mean: 1.38 mmHg  → placebo patients improved slightly
Treatment reduction mean: 11.02 mmHg → medication patients improved strongly
Observed difference:      9.65 mmHg  → treatment was 9.65 mmHg better
Positive value = BP dropped = good outcome for the patient.
Negative value = BP went up = bad outcome.

---

### Step 2 — Descriptive Statistics
mean   → average BP reduction across all 100 patients in the group
std    → how much individual patients varied around that average
std=10 means some patients improved by 1 mmHg, others by 21 mmHg
median → the middle patient's reduction — less affected by extreme values
Q1–Q3  → the middle 50% of patients fell in this range
Treatment Q1–Q3: [3.89, 18.17] → even worst responders improved
Control   Q1–Q3: [-5.23, 5.60] → some control patients got WORSE

---

### Step 3 — Outlier Detection (IQR)
Treatment: 0 outliers → all 100 patients responded within expected range
Control:   5 outliers → 4 unusually large placebo responders + 1 whose BP spiked
Lower fence = Q1 - 1.5×IQR  → below this = unusually bad outcome
Upper fence = Q3 + 1.5×IQR  → above this = unusually good outcome
Outliers are NOT errors — they are real patients with unusual responses.
In a real trial each outlier would be investigated individually.

---

### Step 4 — Bootstrap 95% Confidence Intervals
Treatment CI: [9.04, 13.00]
Control CI:   [-0.57, 3.40]
Meaning: we are 95% confident the TRUE population mean
(all patients who could ever take this drug) lies
in this range — not just our 100-patient sample
CIs do NOT overlap → even accounting for sampling uncertainty
the two groups are clearly different
Width ≈ 4.0 for both → same uncertainty in both groups
because same n=100 and same std≈10

---

### Step 5 — Permutation Test
p-value = 0.0
Meaning: in 10,000 random shuffles of labels, NOT ONCE did
a random group difference reach 9.65 mmHg
95% random range: [-3.09, 3.06]
→ random chance normally produces differences of ±3 mmHg at most
→ our 9.65 is 3x larger than what chance can explain
p < 0.05 → SIGNIFICANT → the medication effect is REAL, not luck

---

### Step 6 — Cohen's d
d = 0.947 → large effect
Meaning: the two group means are 0.95 standard deviations apart
Scale:
d < 0.2 → negligible   (you would not notice in real life)
d = 0.2 → small        (subtle but real)
d = 0.5 → medium       (clearly noticeable)
d = 0.8 → large        (obvious, strong)
d = 0.947 → large      (clinically meaningful)
p-value alone is not enough:
p=0.0 tells you the effect is real
d=0.95 tells you the effect is large enough to matter to patients

---

### Final Conclusion — Reading All Results Together
p-value = 0.0   → effect is definitely real (not luck)
d = 0.947       → effect is large (clinically meaningful)
CI gap = 5.6    → groups are in completely different ranges
observed = 9.65 → medication reduces BP ~10 mmHg more than placebo
Full conclusion:
"The blood pressure medication produces a statistically
significant (p<0.001) and clinically large (d=0.95)
reduction in blood pressure of 9.65 mmHg compared to
placebo. We are 95% confident the true effect lies
between 5.64 and 13.65 mmHg."
This is how results are reported in a real clinical paper.

---

### The 3 Tools — When to Use Each
| Tool | Question it answers | Sensitive to sample size? |
|---|---|---|
| Bootstrap CI | How uncertain is my estimate? | Yes — larger n → narrower CI |
| Permutation test | Is the difference real or luck? | Yes — larger n → smaller p |
| Cohen's d | How big is the effect? | No — pure effect size |

Always report all three together. Never report p-value alone.

In [18]:
# Clinical Trial Analyser 

#1. Generate the trial data
rng= np.random.default_rng(seed=42)
n=100 #100 patients per group

baseline_control= rng.normal(loc=140,scale=15,size=n)
baseline_treatment= rng.normal(loc=140,scale=15,size=n)

outcome_control= baseline_control- rng.normal(2,10,n)
outcome_treatment= baseline_treatment-rng.normal(10,10,n)

reduction_control= baseline_control-outcome_control
reduction_treatment= baseline_treatment-outcome_treatment

observed_diff= np.mean(reduction_treatment)-np.mean(reduction_control)
print(f"Control   reduction — mean: {np.mean(reduction_control):.4f}")
print(f"Treatment reduction — mean: {np.mean(reduction_treatment):.4f}")
print(f"Observed differnce: {observed_diff:4f}")

#Descriptive summary function

def descriptive_summary(arr,label):
    
    mean= np.mean(arr)
    std= np.std(arr,ddof=1)
    median= np.median(arr)
    q1,q3= np.percentile(arr,[25,75])
    print(f"\n{label}")
    print(f"Mean: {mean:.4f}")
    print(f"Std:{std:.4f}")
    print(f"Median:{median:.4f}")
    print(f"Q1-Q3 : {q1:.4f}-{q3:.4f}")
    
    return ({"mean": mean,
             "Std": std, 
             "median": median,
             "Q1"    : q1,
             "Q3"    : q3
            
            })
    

#Outlier Detection

def outlier_detect_iqr(arr,label):
    q1= np.percentile(arr,25)
    q3= np.percentile(arr,75)
    iqr= q3-q1
    lower= q1-1.5 *iqr
    upper= q3+ 1.5*iqr
    
    outlier_mask= (arr<lower)|(arr>upper)
    outlier_values= arr[outlier_mask]
    
    print(f"\nOutliers in {label} : {len(outlier_values)}")
    print(f"Lower fence: {lower:.4f}")
    print(f"Upper fence: {upper:.4f}")
    if len(outlier_values)>0:
        print(f"Values : {outlier_values.round(4)}")
    else:
        print("No outliers detected")
    return {"Count": len(outlier_values),
             "Values": outlier_values}
    
    
    


#Bootstrap_ci
def bootstrap_mean_ci(arr,label,n_bootstrap=10000,confidence=95):
    n=len(arr)
    boot_samples= rng.choice(arr,size=(n_bootstrap,n),replace=True)
    boot_mean= np.mean(boot_samples,axis=1)
    
    alpha= (100-confidence)/2
    ci_lower   = np.percentile(boot_mean,alpha)
    ci_upper= np.percentile(boot_mean,100-alpha)
    
    print(f"\nBootstrap {confidence}% CI -{label}")
    print(f"Mean:{np.mean(arr):.4f}")
    print(f"CI : [{ci_lower:.4f},{ci_upper:.4f}]")
    print(f"Width: {ci_upper-ci_lower:.4f}")
    
    return { "mean":np.mean(arr),
            "ci_lower": ci_lower,
            "ci_upper": ci_upper
    }
    

#Permutation Test
def run_permutation_test(grp_a,grp_b,label_a,label_b,n_permutations=10000):
        observed_diff= np.mean(grp_a)-np.mean(grp_b)
        combined= np.concatenate([grp_a,grp_b])
        n= len(grp_a)
        perm_diff=np.zeros(n_permutations)
        
        for i in range(n_permutations):
            shuffled= rng.permutation(combined)
            perm_diff_value= np.mean(shuffled[:n])-np.mean(shuffled[n:])
            perm_diff[i]= perm_diff_value
            
        p_value= np.mean(np.abs(perm_diff)>= np.abs(observed_diff))
        pct_lower= np.percentile(perm_diff,2.5)
        pct_upper= np.percentile(perm_diff,97.5)
        
        print(f"\nPermutation Test: {label_a}-{label_b}")
        print(f"Observed difference: {observed_diff:.4f}")
        print(f"p-value (two-sided):{p_value}")
        print(f"Significant (alpha<0.05):{p_value<0.05}")
        print(f"95% CI range: [{pct_lower:.4f},{pct_upper:.4f}]")
        
        if (observed_diff>pct_upper )|(observed_diff<pct_lower):
            print(f"  → Observed diff OUTSIDE random range → SIGNIFICANT ✓")
        else:
            print(f"  → Observed diff INSIDE random range → not significant")
            
        return {"observed_diff":observed_diff,"p_value":p_value,"significant":p_value<0.05}
        
        
        
            


#Cohen's d
def compute_cohens_d(group_a, group_b):
    """
    Cohen's d measures HOW BIG the effect is.
    p-value tells you WHETHER the effect is real.
    Cohen's d tells you whether it actually MATTERS.
    """
    mean_diff  = np.mean(group_a) - np.mean(group_b)

    n_a, n_b   = len(group_a), len(group_b)
    var_a      = np.var(group_a, ddof=1)
    var_b      = np.var(group_b, ddof=1)

    # Pooled std — weighted average of both groups' spread
    pooled_std = np.sqrt(((n_a-1)*var_a + (n_b-1)*var_b) / (n_a+n_b-2))

    d = mean_diff / pooled_std

    if   abs(d) < 0.2:  interpretation = "negligible"
    elif abs(d) < 0.5:  interpretation = "small"
    elif abs(d) < 0.8:  interpretation = "medium"
    else:               interpretation = "large"

    print(f"\n  Cohen's d:      {d:.4f}")
    print(f"  Pooled std:     {pooled_std:.4f}")
    print(f"  Interpretation: {interpretation} effect")

    return {'d': d, 'interpretation': interpretation}

def main():
    print("=" * 45)
    print("  CLINICAL TRIAL ANALYSIS REPORT")
    print("  Blood Pressure Medication Study")
    print("=" * 45)

    print("\n── DESCRIPTIVE STATISTICS ──────────────")
    stats_t = descriptive_summary(reduction_treatment, "Treatment Group (Medication)")
    stats_c = descriptive_summary(reduction_control,   "Control Group (Placebo)")

    print("\n── OUTLIER DETECTION (IQR) ─────────────")
    out_t = outlier_detect_iqr(reduction_treatment, "Treatment Group")
    out_c = outlier_detect_iqr(reduction_control,   "Control Group")

    print("\n── BOOTSTRAP CONFIDENCE INTERVALS ──────")
    ci_t = bootstrap_mean_ci(reduction_treatment, "Treatment Group")
    ci_c = bootstrap_mean_ci(reduction_control,   "Control Group")

    print("\n── PERMUTATION TEST ─────────────────────")
    perm = run_permutation_test(
        reduction_treatment, reduction_control,
        "Treatment", "Control"
    )

    print("\n── EFFECT SIZE ──────────────────────────")
    effect = compute_cohens_d(reduction_treatment, reduction_control)

    print("\n── CONCLUSION ───────────────────────────")
    if perm['significant']:
        print(f"  ✅ The medication shows a SIGNIFICANT effect.")
        print(f"  Treatment reduced BP by {perm['observed_diff']:.2f} mmHg")
        print(f"  more than placebo.")
        print(f"  Effect size: {effect['interpretation']} (d = {effect['d']:.3f})")
    else:
        print(f"  ❌ No significant difference found.")
        print(f"  p = {perm['p_value']:.4f}")
    print("=" * 45)

main()


Control   reduction — mean: 1.3766
Treatment reduction — mean: 11.0249
Observed differnce: 9.648258
  CLINICAL TRIAL ANALYSIS REPORT
  Blood Pressure Medication Study

── DESCRIPTIVE STATISTICS ──────────────

Treatment Group (Medication)
Mean: 11.0249
Std:10.1292
Median:12.3364
Q1-Q3 : 3.8889-18.1732

Control Group (Placebo)
Mean: 1.3766
Std:10.2447
Median:0.9961
Q1-Q3 : -5.2344-5.6028

── OUTLIER DETECTION (IQR) ─────────────

Outliers in Treatment Group : 0
Lower fence: -17.5376
Upper fence: 39.5997
No outliers detected

Outliers in Control Group : 5
Lower fence: -21.4902
Upper fence: 21.8586
Values : [ 27.1747  23.2925  31.0507  22.9317 -23.6666]

── BOOTSTRAP CONFIDENCE INTERVALS ──────

Bootstrap 95% CI -Treatment Group
Mean:11.0249
CI : [9.0440,12.9952]
Width: 3.9512

Bootstrap 95% CI -Control Group
Mean:1.3766
CI : [-0.5731,3.3950]
Width: 3.9681

── PERMUTATION TEST ─────────────────────

Permutation Test: Treatment-Control
Observed difference: 9.6483
p-value (two-sided):0.0
Si